[Bug]: Assertion hit in is_valid_now_expr

### What type of bug is this?

Crash

### What subsystems and features are affected?

Query planner

### What happened?

A crash occurred (triggered by an assert hit) when running Sqlsmith.

Link to failed CI job: https://github.com/timescale/timescaledb/actions/runs/3057498072/jobs/4932712809

```
2022-09-15T03:31:16.1260746Z 2022-09-15 03:31:16 Seed 344975
2022-09-15T03:34:43.1938048Z SQLsmith 2014d21
2022-09-15T03:34:43.2047782Z Loading types...done.
2022-09-15T03:34:43.2306297Z Loading tables...done.
2022-09-15T03:34:43.2879888Z Loading columns and constraints...done.
2022-09-15T03:34:43.2901420Z Loading operators...done.
2022-09-15T03:34:43.3066691Z Loading routines...done.
2022-09-15T03:34:45.5051464Z Loading routine parameters...done.
2022-09-15T03:34:45.5095862Z Loading aggregates...done.
2022-09-15T03:34:45.6106131Z Loading aggregate parameters...done.
2022-09-15T03:34:45.6928479Z Generating indexes...done.
2022-09-15T03:34:45.7002430Z 2022-09-15 03:34:45 Seed 11872767
2022-09-15T03:38:06.9761342Z ##[group]Run # wait for in progress coredumps
2022-09-15T03:38:06.9761726Z [36;1m# wait for in progress coredumps[0m
2022-09-15T03:38:06.9761966Z [36;1msleep 10[0m
2022-09-15T03:38:06.9762182Z [36;1mif coredumpctl list; then[0m
2022-09-15T03:38:06.9762460Z [36;1m  echo "::set-output name=coredumps::true"[0m
2022-09-15T03:38:06.9762704Z [36;1m  false[0m
2022-09-15T03:38:06.9762888Z [36;1mfi[0m
2022-09-15T03:38:06.9828689Z shell: /usr/bin/bash -e {0}
2022-09-15T03:38:06.9828912Z env:
2022-09-15T03:38:06.9829103Z   PG_SRC_DIR: pgbuild
2022-09-15T03:38:06.9829335Z   PG_INSTALL_DIR: postgresql
2022-09-15T03:38:06.9829552Z ##[endgroup]
2022-09-15T03:38:17.0038225Z TIME                            PID   UID   GID SIG COREFILE  EXE
2022-09-15T03:38:17.0039308Z Thu 2022-09-15 03:17:26 UTC   12915  1001   121   6 present   /home/runner/postgresql/bin/postgres
2022-09-15T03:38:17.0057214Z ##[error]Process completed with exit code 1.
```

### TimescaleDB version affected

2.9.0-dev / cea1a21c01d76f0e421fcfd86e2ae116f92e1943

### PostgreSQL version used

14.5

### What operating system did you use?

ubuntu-20.04

### What installation method did you use?

Source

### What platform did you run on?

On prem/Self-hosted, Other

### Relevant log output and stack trace

```bash
2022-09-15T03:38:18.1077332Z Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
2022-09-15T03:38:18.7550501Z 50	../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
2022-09-15T03:38:18.7551155Z Core was generated by `postgres: runner smith [local] SELECT                                         '.
2022-09-15T03:38:18.7552129Z Program terminated with signal SIGABRT, Aborted.
2022-09-15T03:38:18.7552507Z #0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
2022-09-15T03:38:18.8230931Z (gdb) #0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
2022-09-15T03:38:18.8235121Z         set = {__val = {4194304, 0, 0, 0, 0, 93993530820496, 
2022-09-15T03:38:18.8235407Z             4611686018427387904, 0, 0, 140316229563626, 0, 93993530820542, 
2022-09-15T03:38:18.8237162Z             128, 0, 140734026272608, 140316223637024}}
2022-09-15T03:38:18.8237663Z         pid = <optimized out>
2022-09-15T03:38:18.8238045Z         tid = <optimized out>
2022-09-15T03:38:18.8238271Z         ret = <optimized out>
2022-09-15T03:38:18.8238534Z #1  0x00007f9dea96b859 in __GI_abort () at abort.c:79
2022-09-15T03:38:18.8304471Z         save_stage = 1
2022-09-15T03:38:18.8304931Z         act = {__sigaction_handler = {sa_handler = 0x7f9deaafebf7, 
2022-09-15T03:38:18.8305496Z             sa_sigaction = 0x7f9deaafebf7}, sa_mask = {__val = {1, 
2022-09-15T03:38:18.8306049Z               140316223995934, 3, 140734026272788, 12, 140316223995938, 2, 
2022-09-15T03:38:18.8306363Z               7149241437714225784, 7305462225721897785, 140734026272896, 
2022-09-15T03:38:18.8306821Z               7291948132566590831, 93993557749424, 2473360997922896384, 
2022-09-15T03:38:18.8307206Z               140734026272880, 259, 19}}, sa_flags = 832884864, 
2022-09-15T03:38:18.8307459Z           sa_restorer = 0x7f9de0bf656f}
2022-09-15T03:38:18.8307724Z         sigs = {__val = {32, 0 <repeats 15 times>}}
2022-09-15T03:38:18.8308007Z #2  0x0000557c934b9758 in ExceptionalCondition (
2022-09-15T03:38:18.8308755Z     conditionName=conditionName@entry=0x7f9de0bf656f "n >= 0 && n < list->length", errorType=errorType@entry=0x7f9de0bf632d "FailedAssertion", 
2022-09-15T03:38:18.8309295Z     fileName=fileName@entry=0x7f9de0bf6478 "/home/runner/postgresql/include/server/nodes/pg_list.h", lineNumber=lineNumber@entry=259) at assert.c:69
2022-09-15T03:38:18.8309650Z No locals.
2022-09-15T03:38:18.8309890Z #3  0x00007f9de0be959f in list_nth_cell (n=<optimized out>, 
2022-09-15T03:38:18.8310149Z     list=<optimized out>)
2022-09-15T03:38:18.8310437Z     at /home/runner/postgresql/include/server/nodes/pg_list.h:259
2022-09-15T03:38:18.8474744Z No locals.
2022-09-15T03:38:18.8475056Z #4  is_valid_now_expr (rtable=<optimized out>, op=<optimized out>, 
2022-09-15T03:38:18.8475339Z     op=<optimized out>)
2022-09-15T03:38:18.8475659Z     at /home/runner/work/timescaledb/timescaledb/src/planner/constify_now.c:53
2022-09-15T03:38:18.8485373Z         flags = <optimized out>
2022-09-15T03:38:18.8485765Z         var = <optimized out>
2022-09-15T03:38:18.8486122Z         rte = <optimized out>
2022-09-15T03:38:18.8488421Z         dim = <optimized out>
2022-09-15T03:38:18.8488760Z         op_inner = <optimized out>
2022-09-15T03:38:18.8489167Z         c = <optimized out>
2022-09-15T03:38:18.8489593Z #5  0x00007f9de0be96cf in ts_constify_now (root=0x7fff31a4d820, 
2022-09-15T03:38:18.8490027Z     rtable=<optimized out>, node=0x557c949eeeb0)
2022-09-15T03:38:18.8490515Z     at /home/runner/work/timescaledb/timescaledb/src/planner/constify_now.c:225
2022-09-15T03:38:18.8490924Z No locals.
2022-09-15T03:38:18.8491262Z #6  0x00007f9de0be5ca1 in preprocess_query (node=0x557c949ef010, 
2022-09-15T03:38:18.8491600Z     context=0x7fff31a4d6b0)
2022-09-15T03:38:18.8491974Z     at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:341
2022-09-15T03:38:18.8493288Z         from = 0x557c949ef010
2022-09-15T03:38:18.8493638Z #7  0x0000557c93271c67 in query_tree_walker (flags=0, context=0x7fff31a4d6b0, 
2022-09-15T03:38:18.8494046Z     walker=0x7f9de0be57b0 <preprocess_query>, query=0x557c94f222b8)
2022-09-15T03:38:18.8494410Z     at nodeFuncs.c:2385
2022-09-15T03:38:18.8496911Z No locals.
2022-09-15T03:38:18.8497218Z #8  query_tree_walker (query=query@entry=0x557c94f222b8, 
2022-09-15T03:38:18.8497591Z     walker=walker@entry=0x7f9de0be57b0 <preprocess_query>, 
2022-09-15T03:38:18.8497980Z     context=context@entry=0x7fff31a4d6b0, flags=flags@entry=0)
2022-09-15T03:38:18.8498314Z     at nodeFuncs.c:2363
2022-09-15T03:38:18.8499621Z         lc = <optimized out>
2022-09-15T03:38:18.8499954Z         lc__state = {l = <optimized out>, i = <optimized out>}
2022-09-15T03:38:18.8500281Z         wc = <optimized out>
2022-09-15T03:38:18.8500632Z #9  0x00007f9de0be5a4c in preprocess_query (node=<optimized out>, 
2022-09-15T03:38:18.8500977Z     context=0x7fff31a4d6b0)
2022-09-15T03:38:18.8501350Z     at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:435
2022-09-15T03:38:18.8505434Z         prev_query = 0x557c94f22088
2022-09-15T03:38:18.8505740Z         hcache = <optimized out>
2022-09-15T03:38:18.8506042Z         rti = <optimized out>
2022-09-15T03:38:18.8506331Z         ret = <optimized out>
2022-09-15T03:38:18.8506615Z         query = <optimized out>
2022-09-15T03:38:18.8506906Z         lc = <optimized out>
2022-09-15T03:38:18.8507256Z #10 0x0000557c93271ac3 in range_table_entry_walker (rte=0x557c94f221a0, 
2022-09-15T03:38:18.8507686Z     walker=0x7f9de0be57b0 <preprocess_query>, context=0x7fff31a4d6b0, flags=0)
2022-09-15T03:38:18.8508044Z     at nodeFuncs.c:2522
2022-09-15T03:38:18.8509095Z No locals.
2022-09-15T03:38:18.8509888Z #11 0x0000557c93271ba8 in range_table_walker (rtable=0x557c949ef248, 
2022-09-15T03:38:18.8510379Z     walker=walker@entry=0x7f9de0be57b0 <preprocess_query>, 
2022-09-15T03:38:18.8510776Z     context=context@entry=0x7fff31a4d6b0, flags=flags@entry=0)
2022-09-15T03:38:18.8511101Z     at nodeFuncs.c:2473
2022-09-15T03:38:18.8513688Z         rte = <optimized out>
2022-09-15T03:38:18.8513998Z         rt__state = {l = 0x557c949ef248, i = 0}
2022-09-15T03:38:18.8514302Z         rt = <optimized out>
2022-09-15T03:38:18.8514657Z #12 0x0000557c93271d5e in query_tree_walker (flags=0, context=0x7fff31a4d6b0, 
2022-09-15T03:38:18.8515060Z     walker=0x7f9de0be57b0 <preprocess_query>, query=0x557c94f22088)
2022-09-15T03:38:18.8515393Z     at nodeFuncs.c:2450
2022-09-15T03:38:18.8676204Z No locals.
2022-09-15T03:38:18.8676712Z #13 0x00007f9de0be5a4c in preprocess_query (node=<optimized out>, 
2022-09-15T03:38:18.8677110Z     context=0x7fff31a4d6b0)
2022-09-15T03:38:18.8677498Z     at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:435
2022-09-15T03:38:18.8684482Z         prev_query = 0x557c94f22088
2022-09-15T03:38:18.8684857Z         hcache = <optimized out>
2022-09-15T03:38:18.8685192Z         rti = <optimized out>
2022-09-15T03:38:18.8685512Z         ret = <optimized out>
2022-09-15T03:38:18.8685811Z         query = <optimized out>
2022-09-15T03:38:18.8686131Z         lc = <optimized out>
2022-09-15T03:38:18.8686479Z #14 0x00007f9de0be602e in timescaledb_planner (parse=0x557c94f22088, 
2022-09-15T03:38:18.8687134Z     query_string=0x557c9471bf20 "select  \n  subq_2.c0 as c0\nfrom \n  (select  \n        sample_1.time as c0\n      from \n        public.conditions_dist as sample_0 tablesample bernoulli (1.6) \n          left join public.sensors as ref_0"..., cursor_opts=2048, bound_params=0x0)
2022-09-15T03:38:18.8687799Z     at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:491
2022-09-15T03:38:18.8840352Z         context = {rootquery = 0x557c94f22088, 
2022-09-15T03:38:18.8840844Z           current_query = 0x557c94f222b8, root = 0x7fff31a4d820, 
2022-09-15T03:38:18.8841263Z           num_distributed_tables = 2}
2022-09-15T03:38:18.8841646Z         glob = {type = T_Invalid, boundParams = 0x0, subplans = 0x0, 
2022-09-15T03:38:18.8842065Z           subroots = 0x0, rewindPlanIDs = 0x0, finalrtable = 0x0, 
2022-09-15T03:38:18.8842480Z           finalrowmarks = 0x0, resultRelations = 0x0, appendRelations = 0x0, 
2022-09-15T03:38:18.8842899Z           relationOids = 0x0, invalItems = 0x0, paramExecTypes = 0x0, 
2022-09-15T03:38:18.8843309Z           lastPHId = 0, lastRowMarkId = 0, lastPlanNodeId = 0, 
2022-09-15T03:38:18.8843698Z           transientPlan = false, dependsOnRole = false, 
2022-09-15T03:38:18.8844090Z           parallelModeOK = false, parallelModeNeeded = false, 
2022-09-15T03:38:18.8844743Z           maxParallelHazard = 0 '\000', partition_directory = 0x0}
2022-09-15T03:38:18.8845152Z         root = {type = T_Invalid, parse = 0x0, glob = 0x7fff31a4d6d0, 
2022-09-15T03:38:18.8845541Z           query_level = 0, parent_root = 0x0, plan_params = 0x0, 
2022-09-15T03:38:18.8845923Z           outer_params = 0x0, simple_rel_array = 0x0, 
2022-09-15T03:38:18.8846292Z           simple_rel_array_size = 0, simple_rte_array = 0x0, 
2022-09-15T03:38:18.8846655Z           append_rel_array = 0x0, all_baserels = 0x0, 
2022-09-15T03:38:18.8847038Z           nullable_baserels = 0x0, join_rel_list = 0x0, join_rel_hash = 0x0, 
2022-09-15T03:38:18.8847434Z           join_rel_level = 0x0, join_cur_level = 0, init_plans = 0x0, 
2022-09-15T03:38:18.8847831Z           cte_plan_ids = 0x0, multiexpr_params = 0x0, eq_classes = 0x0, 
2022-09-15T03:38:18.8848214Z           ec_merging_done = false, canon_pathkeys = 0x0, 
2022-09-15T03:38:18.8848579Z           left_join_clauses = 0x0, right_join_clauses = 0x0, 
2022-09-15T03:38:18.8848932Z           full_join_clauses = 0x0, join_info_list = 0x0, 
2022-09-15T03:38:18.8849280Z           all_result_relids = 0x0, leaf_result_relids = 0x0, 
2022-09-15T03:38:18.8849823Z           append_rel_list = 0x0, row_identity_vars = 0x0, rowMarks = 0x0, 
2022-09-15T03:38:18.8850299Z           placeholder_list = 0x0, fkey_list = 0x0, query_pathkeys = 0x0, 
2022-09-15T03:38:18.8850687Z           group_pathkeys = 0x0, window_pathkeys = 0x0, 
2022-09-15T03:38:18.8851075Z           distinct_pathkeys = 0x0, sort_pathkeys = 0x0, part_schemes = 0x0, 
2022-09-15T03:38:18.8851474Z           initial_rels = 0x0, upper_rels = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 
2022-09-15T03:38:18.8851853Z             0x0}, upper_targets = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, 
2022-09-15T03:38:18.8852238Z           processed_tlist = 0x0, update_colnos = 0x0, grouping_map = 0x0, 
2022-09-15T03:38:18.8852637Z           minmax_aggs = 0x0, planner_cxt = 0x0, total_table_pages = 0, 
2022-09-15T03:38:18.8853032Z           tuple_fraction = 0, limit_tuples = 0, qual_security_level = 0, 
2022-09-15T03:38:18.8853460Z           hasJoinRTEs = false, hasLateralRTEs = false, hasHavingQual = false, 
2022-09-15T03:38:18.8853912Z           hasPseudoConstantQuals = false, hasAlternativeSubPlans = false, 
2022-09-15T03:38:18.8854326Z           hasRecursion = false, agginfos = 0x0, aggtransinfos = 0x0, 
2022-09-15T03:38:18.8854725Z           numOrderedAggs = 0, hasNonPartialAggs = false, 
2022-09-15T03:38:18.8855100Z           hasNonSerialAggs = false, wt_param_id = 0, 
2022-09-15T03:38:18.8855499Z           non_recursive_path = 0x0, curOuterRels = 0x0, curOuterParams = 0x0, 
2022-09-15T03:38:18.8855919Z           isAltSubplan = 0x0, isUsedSubplan = 0x0, join_search_private = 0x0, 
2022-09-15T03:38:18.8856282Z           partColsUpdated = false}
2022-09-15T03:38:18.8856622Z         _save_exception_stack = 0x7fff31a4de30
2022-09-15T03:38:18.8856938Z         _save_context_stack = 0x0
2022-09-15T03:38:18.8857381Z         _local_sigjmp_buf = {{__jmpbuf = {0, -4511251474115938921, 
2022-09-15T03:38:18.8857960Z               93993554788128, 2048, 0, 0, -4511251474520689257, 
2022-09-15T03:38:18.8858399Z               -4494184944948169321}, __mask_was_saved = 0, __saved_mask = {
2022-09-15T03:38:18.8858764Z               __val = {140316057696184, 93993557734208, 140316057696184, 
2022-09-15T03:38:18.8859094Z                 93993555017288, 93993535713625, 20265, 140316057696184, 0, 
2022-09-15T03:38:18.8859417Z                 93993535455869, 1, 93993530903277, 70368744197929, 
2022-09-15T03:38:18.8859741Z                 2473360997922896384, 93993557733512, 93993534009740, 
2022-09-15T03:38:18.8860046Z                 140734026274885}}}}
2022-09-15T03:38:18.8860345Z         _do_rethrow = <optimized out>
2022-09-15T03:38:18.8860657Z         stmt = <optimized out>
2022-09-15T03:38:18.8860960Z         lc = <optimized out>
2022-09-15T03:38:18.8861276Z         reset_fetcher_type = <optimized out>
2022-09-15T03:38:18.8861613Z         reset_baserel_info = <optimized out>
2022-09-15T03:38:18.8861931Z         __func__ = "timescaledb_planner"
2022-09-15T03:38:18.8862420Z #15 0x0000557c93398d0c in pg_plan_query (querytree=0x557c94f22088, 
2022-09-15T03:38:18.8863097Z     query_string=0x557c9471bf20 "select  \n  subq_2.c0 as c0\nfrom \n  (select  \n        sample_1.time as c0\n      from \n        public.conditions_dist as sample_0 tablesample bernoulli (1.6) \n          left join public.sensors as ref_0"..., cursorOptions=2048, boundParams=0x0) at postgres.c:846
2022-09-15T03:38:18.8863741Z         plan = <optimized out>
2022-09-15T03:38:18.8864086Z #16 0x0000557c93398e01 in pg_plan_queries (querytrees=0x557c949ef950, 
2022-09-15T03:38:18.8864756Z     query_string=query_string@entry=0x557c9471bf20 "select  \n  subq_2.c0 as c0\nfrom \n  (select  \n        sample_1.time as c0\n      from \n        public.conditions_dist as sample_0 tablesample bernoulli (1.6) \n          left join public.sensors as ref_0"..., cursorOptions=cursorOptions@entry=2048, 
2022-09-15T03:38:18.8865421Z     boundParams=boundParams@entry=0x0) at postgres.c:938
2022-09-15T03:38:18.8907912Z         query = <optimized out>
2022-09-15T03:38:18.8908311Z         stmt = <optimized out>
2022-09-15T03:38:18.8908792Z         query_list__state = {l = 0x557c949ef950, i = 0}
2022-09-15T03:38:18.8909171Z         stmt_list = 0x0
2022-09-15T03:38:18.8909475Z         query_list = <optimized out>
2022-09-15T03:38:18.8909818Z #17 0x0000557c9339920f in exec_simple_query (
2022-09-15T03:38:18.8910402Z     query_string=0x557c9471bf20 "select  \n  subq_2.c0 as c0\nfrom \n  (select  \n        sample_1.time as c0\n      from \n        public.conditions_dist as sample_0 tablesample bernoulli (1.6) \n          left join public.sensors as ref_0"...) at postgres.c:1132
2022-09-15T03:38:18.8934222Z         snapshot_set = true
2022-09-15T03:38:18.8934611Z         per_parsetree_context = 0x0
2022-09-15T03:38:18.8934987Z         plantree_list = <optimized out>
2022-09-15T03:38:18.8935319Z         parsetree = <optimized out>
2022-09-15T03:38:18.8935647Z         commandTag = CMDTAG_SELECT
2022-09-15T03:38:18.8936014Z         qc = {commandTag = CMDTAG_UNKNOWN, nprocessed = 93991064305674}
2022-09-15T03:38:18.8936379Z         querytree_list = <optimized out>
2022-09-15T03:38:18.8936702Z         portal = <optimized out>
2022-09-15T03:38:18.8937022Z         receiver = <optimized out>
2022-09-15T03:38:18.8937306Z         format = 0
2022-09-15T03:38:18.8937631Z         parsetree_item__state = {l = 0x557c94f21f18, i = 0}
2022-09-15T03:38:18.8937948Z         dest = DestRemote
2022-09-15T03:38:18.8938240Z         oldcontext = 0x557c947bc8e0
2022-09-15T03:38:18.8938551Z         parsetree_list = 0x557c94f21f18
2022-09-15T03:38:18.8938875Z         parsetree_item = <optimized out>
2022-09-15T03:38:18.8939192Z         save_log_statement_stats = false
2022-09-15T03:38:18.8939485Z         was_logged = false
2022-09-15T03:38:18.8939778Z         use_implicit_block = false
2022-09-15T03:38:18.8940191Z         msec_str = "\000\000\000\000\000\000\000\000\000\372\177\"\271$S\"\000\000\000\000\000\000\000\000+\323L\223|U\000"
2022-09-15T03:38:18.8940569Z         __func__ = "exec_simple_query"
2022-09-15T03:38:18.8940921Z #18 0x0000557c9339ae36 in PostgresMain (argc=argc@entry=1, 
2022-09-15T03:38:18.8941311Z     argv=argv@entry=0x7fff31a4e010, dbname=<optimized out>, 
2022-09-15T03:38:18.8941688Z     username=<optimized out>) at postgres.c:4496
2022-09-15T03:38:18.8944395Z         query_string = 0x557c9471bf20 "select  \n  subq_2.c0 as c0\nfrom \n  (select  \n        sample_1.time as c0\n      from \n        public.conditions_dist as sample_0 tablesample bernoulli (1.6) \n          left join public.sensors as ref_0"...
2022-09-15T03:38:18.8945029Z         firstchar = <optimized out>
2022-09-15T03:38:18.8945329Z         input_message = {
2022-09-15T03:38:18.8945888Z           data = 0x557c9471bf20 "select  \n  subq_2.c0 as c0\nfrom \n  (select  \n        sample_1.time as c0\n      from \n        public.conditions_dist as sample_0 tablesample bernoulli (1.6) \n          left join public.sensors as ref_0"..., len = 2780, maxlen = 4096, cursor = 2780}
2022-09-15T03:38:18.8946661Z         local_sigjmp_buf = {{__jmpbuf = {0, -4511251474153687657, 1587412965, 
2022-09-15T03:38:18.8947117Z               93993554978280, 3, 93993538758312, -4511251474170464873, 
2022-09-15T03:38:18.8947549Z               -7754189665734183529}, __mask_was_saved = 1, __saved_mask = {
2022-09-15T03:38:18.8947898Z               __val = {4194304, 140734026289105, 0, 0, 140734026276848, 
2022-09-15T03:38:18.8948225Z                 140316222187200, 2473360997922896384, 206158430232, 
2022-09-15T03:38:18.8948547Z                 93993535857443, 206158430240, 140734026276768, 
2022-09-15T03:38:18.8948869Z                 140734026276576, 2473360997922896384, 93993554951872, 0, 6}}}}
2022-09-15T03:38:18.8949191Z         send_ready_for_query = false
2022-09-15T03:38:18.8949533Z         idle_in_transaction_timeout_enabled = false
2022-09-15T03:38:18.8949884Z         idle_session_timeout_enabled = false
2022-09-15T03:38:18.8950202Z         __func__ = "PostgresMain"
2022-09-15T03:38:18.8950549Z #19 0x0000557c93308573 in BackendRun (port=<optimized out>, 
2022-09-15T03:38:18.8950916Z     port=<optimized out>) at postmaster.c:4530
2022-09-15T03:38:18.8951413Z         av = {0x557c9351f7b2 "postgres", 0x0}
2022-09-15T03:38:18.8951772Z         ac = 1
2022-09-15T03:38:18.8952082Z         av = {<optimized out>, <optimized out>}
2022-09-15T03:38:18.8952399Z         ac = <optimized out>
2022-09-15T03:38:18.8952762Z #20 BackendStartup (port=<optimized out>) at postmaster.c:4252
2022-09-15T03:38:18.8953106Z         bn = <optimized out>
2022-09-15T03:38:18.8953399Z         pid = <optimized out>
2022-09-15T03:38:18.8953701Z         bn = <optimized out>
2022-09-15T03:38:18.8953981Z         pid = <optimized out>
2022-09-15T03:38:18.8954281Z         __func__ = "BackendStartup"
2022-09-15T03:38:18.8954606Z         __errno_location = <optimized out>
2022-09-15T03:38:18.8954938Z         __errno_location = <optimized out>
2022-09-15T03:38:18.8955255Z         save_errno = <optimized out>
2022-09-15T03:38:18.8955572Z         __errno_location = <optimized out>
2022-09-15T03:38:18.8955887Z         __errno_location = <optimized out>
2022-09-15T03:38:18.8956211Z #21 ServerLoop () at postmaster.c:1745
2022-09-15T03:38:18.8968054Z         port = <optimized out>
2022-09-15T03:38:18.8968423Z         i = <optimized out>
2022-09-15T03:38:18.8968793Z         rmask = {fds_bits = {128, 0 <repeats 15 times>}}
2022-09-15T03:38:18.8969134Z         selres = <optimized out>
2022-09-15T03:38:18.8969439Z         now = <optimized out>
2022-09-15T03:38:18.8969765Z         readmask = {fds_bits = {224, 0 <repeats 15 times>}}
2022-09-15T03:38:18.8970080Z         nSockets = 8
2022-09-15T03:38:18.8970388Z         last_lockfile_recheck_time = 1663211802
2022-09-15T03:38:18.8970703Z         last_touch_time = 1663211020
2022-09-15T03:38:18.8971001Z         __func__ = "ServerLoop"
2022-09-15T03:38:18.8971348Z #22 0x0000557c933094f9 in PostmasterMain (argc=argc@entry=6, 
2022-09-15T03:38:18.8971712Z     argv=argv@entry=0x557c94713cd0) at postmaster.c:1417
2022-09-15T03:38:18.8972054Z         opt = <optimized out>
2022-09-15T03:38:18.8972344Z         status = <optimized out>
2022-09-15T03:38:18.8972659Z         userDoption = <optimized out>
2022-09-15T03:38:18.8972977Z         listen_addr_saved = true
2022-09-15T03:38:18.8973272Z         i = <optimized out>
2022-09-15T03:38:18.8973597Z         output_config_variable = <optimized out>
2022-09-15T03:38:18.8973930Z         __func__ = "PostmasterMain"
2022-09-15T03:38:18.8974288Z #23 0x0000557c93040712 in main (argc=6, argv=0x557c94713cd0) at main.c:209
2022-09-15T03:38:18.8974647Z         do_check_root = <optimized out>
2022-09-15T03:38:18.8974936Z (gdb) quit
2022-09-15T03:38:18.9396807Z            PID: 12915 (postgres)
2022-09-15T03:38:18.9397310Z            UID: 1001 (runner)
2022-09-15T03:38:18.9397910Z            GID: 121 (docker)
2022-09-15T03:38:18.9398264Z         Signal: 6 (ABRT)
2022-09-15T03:38:18.9398880Z      Timestamp: Thu 2022-09-15 03:17:25 UTC (20min ago)
2022-09-15T03:38:18.9399360Z   Command Line: postgres: runner smith [local] SELECT
2022-09-15T03:38:18.9400230Z     Executable: /home/runner/postgresql/bin/postgres
2022-09-15T03:38:18.9400812Z  Control Group: /system.slice/runner-provisioner.service
2022-09-15T03:38:18.9401341Z           Unit: runner-provisioner.service
2022-09-15T03:38:18.9401717Z          Slice: system.slice
2022-09-15T03:38:18.9402071Z        Boot ID: 78d59b90bb1945b0861984c0bd0ccb4e
2022-09-15T03:38:18.9402412Z     Machine ID: 98a04e6dcf7d48a99624005a7551a332
2022-09-15T03:38:18.9402771Z       Hostname: fv-az212-303
2022-09-15T03:38:18.9403241Z        Storage: /var/lib/systemd/coredump/core.postgres.1001.78d59b90bb1945b0861984c0bd0ccb4e.12915.1663211845000000000000.lz4
2022-09-15T03:38:18.9403738Z        Message: Process 12915 (postgres) of user 1001 dumped core.
2022-09-15T03:38:18.9404062Z                 
2022-09-15T03:38:18.9404353Z                 Stack trace of thread 12915:
2022-09-15T03:38:18.9404703Z                 #0  0x00007f9dea98c00b __GI_raise (libc.so.6 + 0x4300b)
2022-09-15T03:38:18.9405078Z                 #1  0x00007f9dea96b859 __GI_abort (libc.so.6 + 0x22859)
2022-09-15T03:38:18.9405715Z                 #2  0x0000557c934b9758 n/a (/home/runner/postgresql/bin/postgres + 0x5d2758)
```


### How can we reproduce the bug?

```bash
One could possibly run Sqlsmith with seed 11872767. 

However, due to #4708 I was not able to reproduce this issue on my local system so far.
```

--- a/src/planner/constify_now.c
+++ b/src/planner/constify_now.c
@@ -83,6 +83,8 @@ is_valid_now_expr(OpExpr *op, List *rtable)
 		if (!IsA(tle->expr, Var))
 			return false;
 		var = castNode(Var, tle->expr);
+		if (var->varlevelsup != 0)
+			return false;
 		rte = list_nth(rte->subquery->rtable, var->varno - 1);
 	}
 

--- a/tsl/test/shared/expected/constify_now-12.out
+++ b/tsl/test/shared/expected/constify_now-12.out
@@ -491,3 +491,14 @@ QUERY PLAN
 
 DROP TABLE now_view_test CASCADE;
 NOTICE:  drop cascades to view now_view
+-- #4709
+-- test queries with constraints involving columns from different nesting levels
+SELECT * FROM
+  (SELECT * FROM metrics m1 LIMIT 1) m1
+  INNER JOIN (SELECT * FROM metrics m2 LIMIT 1) m2 ON true,
+  LATERAL (SELECT m2.time FROM devices LIMIT 1) as subq_1
+WHERE subq_1.time > m1.time;
+ time | device_id | v0 | v1 | v2 | v3 | time | device_id | v0 | v1 | v2 | v3 | time 
+------+-----------+----+----+----+----+------+-----------+----+----+----+----+------
+(0 rows)
+

--- a/tsl/test/shared/expected/constify_now-13.out
+++ b/tsl/test/shared/expected/constify_now-13.out
@@ -491,3 +491,14 @@ QUERY PLAN
 
 DROP TABLE now_view_test CASCADE;
 NOTICE:  drop cascades to view now_view
+-- #4709
+-- test queries with constraints involving columns from different nesting levels
+SELECT * FROM
+  (SELECT * FROM metrics m1 LIMIT 1) m1
+  INNER JOIN (SELECT * FROM metrics m2 LIMIT 1) m2 ON true,
+  LATERAL (SELECT m2.time FROM devices LIMIT 1) as subq_1
+WHERE subq_1.time > m1.time;
+ time | device_id | v0 | v1 | v2 | v3 | time | device_id | v0 | v1 | v2 | v3 | time 
+------+-----------+----+----+----+----+------+-----------+----+----+----+----+------
+(0 rows)
+

--- a/tsl/test/shared/expected/constify_now-14.out
+++ b/tsl/test/shared/expected/constify_now-14.out
@@ -488,3 +488,14 @@ QUERY PLAN
 
 DROP TABLE now_view_test CASCADE;
 NOTICE:  drop cascades to view now_view
+-- #4709
+-- test queries with constraints involving columns from different nesting levels
+SELECT * FROM
+  (SELECT * FROM metrics m1 LIMIT 1) m1
+  INNER JOIN (SELECT * FROM metrics m2 LIMIT 1) m2 ON true,
+  LATERAL (SELECT m2.time FROM devices LIMIT 1) as subq_1
+WHERE subq_1.time > m1.time;
+ time | device_id | v0 | v1 | v2 | v3 | time | device_id | v0 | v1 | v2 | v3 | time 
+------+-----------+----+----+----+----+------+-----------+----+----+----+----+------
+(0 rows)
+

--- a/tsl/test/shared/sql/constify_now.sql.in
+++ b/tsl/test/shared/sql/constify_now.sql.in
@@ -170,4 +170,11 @@ CREATE VIEW now_view AS SELECT time, device, avg(value) from now_view_test GROUP
 :PREFIX SELECT * FROM now_view WHERE time > now() - '168h'::interval;
 DROP TABLE now_view_test CASCADE;
 
+-- #4709
+-- test queries with constraints involving columns from different nesting levels
+SELECT * FROM
+  (SELECT * FROM metrics m1 LIMIT 1) m1
+  INNER JOIN (SELECT * FROM metrics m2 LIMIT 1) m2 ON true,
+  LATERAL (SELECT m2.time FROM devices LIMIT 1) as subq_1
+WHERE subq_1.time > m1.time;
 

